### PR TITLE
board: add cy8ckit 062 pioneer

### PR DESCRIPTION
**Pull request by DaWei8823**
Issue: #60013 

-Added support for the [CY8CKIT-062S4 Pioneer kit](https://www.infineon.com/cms/en/product/evaluation-boards/cy8ckit-062s4/?redirId=VL1508&utm_medium=referral&utm_source=cypress&utm_campaign=202110_globe_en_all_integration-dev_kit)

- Fixed issues with adc registers based on [CY8C62x4 Registers Technical
Reference Manual (TRM)](https://www.infineon.com/dgdl/Infineon-PSOC_6_MCU_CY8C61X4CY8C62X4_REGISTERS_TECHNICAL_REFERENCE_MANUAL_(TRM)_PSOC_61_PSOC_62_MCU-AdditionalTechnicalInformation-v03_00-EN.pdf?fileId=8ac78c8c7d0d8da4017d0fb34f0627a7), which were throwing dts errors

- removed `/delete-node/ &scb3;` in dts/arm/infineon/psoc6/mpns/CY8C6244LQI_S4D92.dtsi, which seemed erroneous and was throwing dts error

- removed flash1 from dts/arm/infineon/psoc6/psoc6_04/psoc6_04.dtsi, which does not exist in [SOC](https://www.infineon.com/dgdl/Infineon-PSoC_6_MCU_CY8C62X4-DataSheet-v12_00-EN.pdf?fileId=8ac78c8c7ddc01d7017ddd026d585901) and was throwiong dts error  

- Adding `west flash` support with `cy8c6xxa` target pn pyocd which works for flashing this board 

Tested with `west build -b cy8ckit_062_pioneer_m4  --pristine &&  west flash` in [hello_world](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/hello_world) and [blinky project](https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/basic/blinky). Verified that correct console message was printed out on serial terminal and LED blinked.

### CODE REVIEW

File: boards/arm/cy8ckit_062_pioneer/cy8ckit_062_pioneer_m4.dts
Comment (cfriedt) [inline boards/arm/cy8ckit_062_pioneer/cy8ckit_062_pioneer_m4.dts:1]: C-style block comments for compliance please

File: dts/arm/infineon/psoc6/psoc6_04/psoc6_04.dtsi
Comment (cfriedt) [inline dts/arm/infineon/psoc6/psoc6_04/psoc6_04.dtsi:39]: This might break some other use cases. Is there a way to push this node to a separate dtsi?

File: dts/arm/infineon/psoc6/mpns/CY8C6244LQI_S4D92.dtsi
Comment (cfriedt) [inline dts/arm/infineon/psoc6/mpns/CY8C6244LQI_S4D92.dtsi:11]: Looks suspicious to me, but I'm not super familiar with this platform

**Review by cfriedt — COMMENTED** [review]
A couple of comments

**Comment by cfriedt** [comment]
> removed /delete-node/ &scb3; in dts/arm/infineon/psoc6/mpns/CY8C6244LQI_S4D92.dtsi, which seemed erroneous and was throwing dts error
>
> removed flash1 from dts/arm/infineon/psoc6/psoc6_04/psoc6_04.dtsi, which does not exist in SOC and was throwiong dts error

Are you able to verify that other boards based on this soc / family / series are still able to build fine?

File: dts/arm/infineon/psoc6/mpns/CY8C6244LQI_S4D92.dtsi
Comment (npal-cy) [inline dts/arm/infineon/psoc6/mpns/CY8C6244LQI_S4D92.dtsi:11]: CY8C6244LQI_S4D92 does not have scb3, that why we have  '/delete-node/ &scb3;'. What kind of problem do you have if do not remove it..?

File: boards/arm/cy8ckit_062_pioneer/Kconfig.board
Comment (ifyall) [inline boards/arm/cy8ckit_062_pioneer/Kconfig.board:5]: There are multiple BLE kits. I suggest:
```suggestion
	bool "PSoC6 S4 BLE Pioneer Kit"
```

File: boards/arm/cy8ckit_062_pioneer/cy8ckit_062_pioneer_m4.dts
Comment (ifyall) [inline boards/arm/cy8ckit_062_pioneer/cy8ckit_062_pioneer_m4.dts:8]:
```suggestion
	model = "Infineon PSoC 62S4 Pioneer Kit";
```
When written in text, there is always a space between PSoC and any numerical designation. Also, there are lots of PSoC 6 Pioneer Kits, so adding the full name helps set expectations.

**Review by ifyall — CHANGES_REQUESTED** [review]
Overall comment is that the board directory should be cy8ckit-062s4, rather than cy8ckit_062_pioneer. If we don't change it, a user could be confused as there are multiple Pioneer boards with a PSoC 62 on them. This removes any ambiguity.

File: dts/arm/infineon/psoc6/mpns/CY8C6244LQI_S4D92.dtsi
Comment (DaWei8823) [reply dts/arm/infineon/psoc6/mpns/CY8C6244LQI_S4D92.dtsi:11]: I was getting an error on `/delete-node/ &scb3;` because scb3 was not defined in the first place anywhere in the device tree files for this chip.

File: dts/arm/infineon/psoc6/psoc6_04/psoc6_04.dtsi
Comment (DaWei8823) [reply dts/arm/infineon/psoc6/psoc6_04/psoc6_04.dtsi:39]: I can delete it in the board device tree.  The size for this flash is defined as 0 so I was getting an error. It also does not exist according to the vendor docs so that's why I deleted it here. But  I'm fine to delete it in board specific dts!

File: boards/arm/cy8ckit_062_pioneer/Kconfig.board
Comment (DaWei8823) [reply boards/arm/cy8ckit_062_pioneer/Kconfig.board:5]: ahh sorry I put the wrong name here completely. It's not a ble kit at all. I'm changing to what's on the box exactly.

**Comment by DaWei8823** [comment]
> Overall comment is that the board directory should be cy8ckit-062s4, rather than cy8ckit_062_pioneer. If we don't change it, a user could be confused as there are multiple Pioneer boards with a PSoC 62 on them. This removes any ambiguity.

makes sense.I am changing it to `cy8ckit_062s4` (same as suggested but underscore instead of dash). I am changing board name to cy8ckit_062s4_m4


> Are you able to verify that other boards based on this soc / family / series are still able to build fine?

I've tested with 

cy8ckit_062_ble_m0
cy8ckit_062_ble_m4
cy8ckit_062_wifi_bt_m0
cy8ckit_062_wifi_bt_m4
cy8ckit_062s4_m4
cy8cproto_062_4343w
cy8cproto_063_ble
(all the cy8c boars)

they build fine with hello world project.

@cfriedt @ifyall  @npal-cy thanks for the review!

File: boards/arm/cy8ckit_062s4/cy8ckit_062s4_m4.yaml
Comment (npal-cy) [inline boards/arm/cy8ckit_062s4/cy8ckit_062s4_m4.yaml:13]: please remove '- xtools' it is obsolete in Zephyr.

File: dts/arm/infineon/psoc6/mpns/CY8C6244LQI_S4D92.dtsi
Comment (npal-cy) [reply dts/arm/infineon/psoc6/mpns/CY8C6244LQI_S4D92.dtsi:11]: ok. checked this. All, SOC_DIE_PSOC6_04 does not have scb3. so we do not have scb3 in dts\arm\infineon\psoc6\psoc6_04\psoc6_04.dtsi and we do not need `/delete-node/ &scb3;` in MPNs dtsi for SOC_DIE_PSOC6_04.

i will file the ticket to remove `/delete-node/ &scb3;` for all mpns of SOC_DIE_PSOC6_04

File: boards/arm/cy8ckit_062s4/cy8ckit_062s4_m4.dts
Comment (npal-cy) [inline boards/arm/cy8ckit_062s4/cy8ckit_062s4_m4.dts:9]: please move this to dts/arm/infineon/psoc6/mpns/CY8C6244LQI_S4D92.dtsi

File: soc/arm/infineon_cat1/psoc6/Kconfig.defconfig.soc.psoc6_02
Comment (npal-cy) [inline soc/arm/infineon_cat1/psoc6/Kconfig.defconfig.soc.psoc6_02:46]: remove this, you added it in soc/arm/infineon_cat1/psoc6/Kconfig.defconfig.soc.psoc6_04

File: soc/arm/infineon_cat1/psoc6/Kconfig.soc.psoc6_02
Comment (npal-cy) [inline soc/arm/infineon_cat1/psoc6/Kconfig.soc.psoc6_02:175]: move this to soc/arm/infineon_cat1/psoc6/Kconfig.soc.psoc6_04 (need to create)

Comment (npal-cy) [reply soc/arm/infineon_cat1/psoc6/Kconfig.soc.psoc6_02:175]: please use Kconfig.soc.psoc6_04 and Kconfig.defconfig.soc.psoc6_04 from attached archive ([psoc6.zip](https://github.com/zephyrproject-rtos/zephyr/files/12076883/psoc6.zip)
). It has defined all NPMs for psoc6_04

**Comment by npal-cy** [comment]
looks like doc\index.rst is missed.

**Review by npal-cy — CHANGES_REQUESTED** [review]

File: soc/arm/infineon_cat1/psoc6/Kconfig.soc.psoc6_02
Comment (DaWei8823) [reply soc/arm/infineon_cat1/psoc6/Kconfig.soc.psoc6_02:175]: done but I have different values for `NUM_IRQS` based on [datasheet](https://www.infineon.com/dgdl/Infineon-PSoC_6_MCU_CY8C62X4-DataSheet-v12_00-EN.pdf?fileId=8ac78c8c7ddc01d7017ddd026d585901)

I see 175 for cm4 and 16 for cm0 (8 hw and 8 sw irqs)

File: soc/arm/infineon_cat1/psoc6/Kconfig.defconfig.soc.psoc6_04
Comment (ifyall) [inline soc/arm/infineon_cat1/psoc6/Kconfig.defconfig.soc.psoc6_04:35]: This line is a duplicate of line 16. Need to remove.

**Review by ifyall — CHANGES_REQUESTED** [review]
One minor change needed from me.

**Comment by DaWei8823** [comment]
> looks like doc\index.rst is missed.

@npal-cy 

Is there guidance on writing these? Or should I just take inspiration from existing examples?

File: soc/arm/infineon_cat1/psoc6/Kconfig.defconfig.soc.psoc6_04
Comment (DaWei8823) [reply soc/arm/infineon_cat1/psoc6/Kconfig.defconfig.soc.psoc6_04]: done

**Comment by npal-cy** [comment]
> looks like doc\index.rst is missed.

you can start cy8cproto_062_4343w as reference.
more info for added bard you can get here: https://github.com/Infineon/TARGET_CY8CKIT-062S4

**Review by ifyall — DISMISSED** [review]

**Comment by DaWei8823** [comment]
@npal-cy I've add an index.rst

**Review by npal-cy — DISMISSED** [review]

**Comment by cfriedt** [comment]
A couple of minor fixes to the documentation, otherwise LGTM

**Comment by DaWei8823** [comment]
@npal-cy  @ifyall can you please reapprove. all changes since last approval were to doc format/fixing typos

File: boards/arm/cy8ckit_062s4/doc/index.rst
Comment (ifyall) [inline boards/arm/cy8ckit_062s4/doc/index.rst:86]: @DaWei8823, we recently updated all the existing boards with instructions for acquiring the OpenOCD that supports the PSoC 6 devices in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/60703. Can you add a similar note in this section so the customer can get west flash working?

Thanks,

Ian

**Review by ifyall — CHANGES_REQUESTED** [review]

File: boards/arm/cy8ckit_062s4/doc/index.rst
Comment (DaWei8823) [reply boards/arm/cy8ckit_062s4/doc/index.rst:86]: done!

**Review by ifyall — APPROVED** [review]

**Comment by DaWei8823** [comment]
I see [this workflow](https://github.com/zephyrproject-rtos/zephyr/actions/runs/5660328219/job/15335792528?pr=60413) is failing because I'm between these two commits to change path of esp32 soc directory and update the maintainers file. It has nothing to do with this PR

- https://github.com/zephyrproject-rtos/zephyr/commit/6b57b3b786be8558e7e310b849e0b97bd1eae45c
- https://github.com/zephyrproject-rtos/zephyr/commit/5b432ecda42ee9f5c1192d99685356d0c983592b

**Comment by cfriedt** [comment]
> I see [this workflow](https://github.com/zephyrproject-rtos/zephyr/actions/runs/5660328219/job/15335792528?pr=60413) is failing because I'm between these two commits to change path of esp32 soc directory and update the maintainers file. It has nothing to do with this PR
> 
> * [6b57b3b](https://github.com/zephyrproject-rtos/zephyr/commit/6b57b3b786be8558e7e310b849e0b97bd1eae45c)
> * [5b432ec](https://github.com/zephyrproject-rtos/zephyr/commit/5b432ecda42ee9f5c1192d99685356d0c983592b)

@DaWei8823 - can you fetch / rebase on `origin/main`?

**Comment by DaWei8823** [comment]
> > I see [this workflow](https://github.com/zephyrproject-rtos/zephyr/actions/runs/5660328219/job/15335792528?pr=60413) is failing because I'm between these two commits to change path of esp32 soc directory and update the maintainers file. It has nothing to do with this PR
> > 
> > * [6b57b3b](https://github.com/zephyrproject-rtos/zephyr/commit/6b57b3b786be8558e7e310b849e0b97bd1eae45c)
> > * [5b432ec](https://github.com/zephyrproject-rtos/zephyr/commit/5b432ecda42ee9f5c1192d99685356d0c983592b)
> 
> @DaWei8823 - can you fetch / rebase on `origin/main`?

yeah but then I assume I'll have to get the approvals again?

**Comment by cfriedt** [comment]
@DaWei8823 - if we can get people to close the issues in CI, then you probably don't need to get approvals again.

**Comment by DaWei8823** [comment]
> 

@cfriedt now it seems to be failing on an unrelated [Twister issue](https://github.com/zephyrproject-rtos/zephyr/actions/runs/5663117912/job/15344338182?pr=60413) and Twister seems to be  [failing on mainline for last 8 hrs or so](https://github.com/zephyrproject-rtos/zephyr/actions/workflows/twister.yaml?query=branch%3Amain)

**Comment by DaWei8823** [comment]
@npal-cy  can you please reapprove now that i've rebased passed all CI problems?

**Review by cfriedt — APPROVED** [review]

File: boards/arm/cy8ckit_062s4/doc/index.rst
Comment (kartben) [inline boards/arm/cy8ckit_062s4/doc/index.rst:8]: 62S$ --> 62S4
not a big deal but @DaWei8823, if you ever get a chance to submit a fix... :)

Comment (DaWei8823) [reply boards/arm/cy8ckit_062s4/doc/index.rst:8]: @kartben thanks for the catch, I have this PR https://github.com/zephyrproject-rtos/zephyr/pull/61456

**Pull request merged**
